File: typescript/src/parsers/outputParser.ts
/**
 * Parser for Terraform output blocks.
 * Extracts output value declarations with their expressions and metadata.
 */

import { HclBlock, OutputBlock } from '../types/blocks';
import { parseBlockBody } from '../utils/bodyParser';
import { literalBoolean, literalString } from '../utils/valueHelpers';

/**
 * Parser for Terraform output definition blocks.
 *
 * @example
 * ```hcl
 * output "instance_ip" {
 *   description = "The public IP of the instance"
 *   value       = aws_instance.web.public_ip
 *   sensitive   = false
 * }
 * ```
 */
export class OutputParser {
    /**
     * Parses an output block into a structured OutputBlock.
     * @param block - The raw HCL block to parse
     * @returns Parsed OutputBlock with all extracted fields
     */
    parse(block: HclBlock): OutputBlock {
        const name = block.labels[0] || 'unknown';
        const parsed = parseBlockBody(block.body);

        const description = literalString(parsed.attributes.description) ?? parsed.attributes.description?.raw;
        const value = parsed.attributes.value;
        const sensitive = literalBoolean(parsed.attributes.sensitive);

        return {
            name,
            description,
            value,
            sensitive,
            raw: block.raw,
            source: block.source
        };
    }
}

File: python/src/parse_hcl/parsers/output_parser.py
from __future__ import annotations

from typing import Dict

from ..types import HclBlock
from ..utils.common.value_helpers import literal_boolean, literal_string
from ..utils.parser.body_parser import parse_block_body


class OutputParser:
    def parse(self, block: HclBlock) -> Dict[str, object]:
        name = block["labels"][0] if block["labels"] else "unknown"
        parsed = parse_block_body(block["body"])

        description_val = parsed["attributes"].get("description")  # type: ignore[index]
        description = literal_string(description_val) or (description_val.get("raw") if isinstance(description_val, dict) else None)
        value = parsed["attributes"].get("value")  # type: ignore[index]
        sensitive = literal_boolean(parsed["attributes"].get("sensitive"))  # type: ignore[index]

        return {
            "name": name,
            "description": description,
            "value": value,
            "sensitive": sensitive,
            "raw": block["raw"],
            "source": block["source"],
        }
